File: 프로그래머스/lv2/92342. 양궁대회/양궁대회.py

def cal_score(rinfo, ainfo):
    rscore = 0
    ascore = 0
    for idx in range(len(rinfo)):
        if rinfo[idx] == ainfo[idx] == 0:
            continue
        elif rinfo[idx]>ainfo[idx]:
            rscore+=(10-idx)
        else:
            ascore+=(10-idx)
        
    return rscore - ascore

def better(info2, info1):
    r_info2 = list(reversed(info2))
    r_info1 = list(reversed(info1))
    for idx in range(len(r_info2)):
        if r_info2[idx]>r_info1[idx]:
            return True
        elif r_info1[idx]>r_info2[idx]:
            return False
    return False

def search(idx, n, rinfo, ainfo):
    if idx==11 or n == 0:
        rinfo[-1] = n
        diff = cal_score(rinfo, ainfo)
        return diff, rinfo[:]
    
    # 안 쏘고 넘기는 경우
    diff1, info1 = search(idx+1, n, rinfo, ainfo)
    
    # 화살 쏘는 경우 (복구 해야함)
    if n > ainfo[idx]:
        rinfo[idx] = ainfo[idx]+1
        diff2, info2 = search(idx+1, n - rinfo[idx], rinfo, ainfo)
        rinfo[idx] = 0
        if (diff2>diff1) or ((diff2 == diff1) and (better(info2, info1))):
            return diff2, info2
            
    return diff1, info1
    


def solution(n, info):
    diff, answer = search(0, n, [0]*11, info)
    if diff<=0:
        return [-1]
    return answer

File: 프로그래머스/2/92342. 양궁대회/양궁대회.js
function calScore(rinfo, ainfo) {
  let rscore = 0;
  let ascore = 0;

  for (let i = 0; i < 11; i++) {
    if (rinfo[i] === 0 && ainfo[i] === 0) continue;
    if (rinfo[i] > ainfo[i]) rscore += 10 - i;
    else ascore += 10 - i;
  }

  return rscore - ascore;
}

function better(info2, info1) {
  const rInfo2 = [...info2].reverse();
  const rInfo1 = [...info1].reverse();

  for (let i = 0; i < 11; i++) {
    if (rInfo2[i] > rInfo1[i]) return true;
    if (rInfo1[i] > rInfo2[i]) return false;
  }
  return false;
}

function search(idx, n, rinfo, ainfo) {
  if (idx === 11 || n === 0) {
    rinfo[10] += n; // 남은 화살은 0점에 몰아줌
    const diff = calScore(rinfo, ainfo);
    const result = [...rinfo];
    rinfo[10] -= n; // 복구
    return [diff, result];
  }

  // ① 안 쏘는 경우
  const [diff1, info1] = search(idx + 1, n, rinfo, ainfo);

  // ② 쏘는 경우
  let bestDiff = diff1;
  let bestInfo = info1;

  if (n > ainfo[idx]) {
    rinfo[idx] = ainfo[idx] + 1;
    const [diff2, info2] = search(idx + 1, n - rinfo[idx], rinfo, ainfo);
    rinfo[idx] = 0; // 복구

    if (diff2 > bestDiff || (diff2 === bestDiff && better(info2, bestInfo))) {
      bestDiff = diff2;
      bestInfo = info2;
    }
  }

  return [bestDiff, bestInfo];
}

function solution(n, info) {
  const [diff, answer] = search(0, n, Array(11).fill(0), info);
  if (diff <= 0) return [-1];
  return answer;
}
